### ISSUE TITLE
What loss to choose?

### ISSUE THREAD

**Issue by GiorgioBarnabo** [KevinMusgrave/pytorch-metric-learning#504]
Hi everybody,

thanks a lot for this amazing library. It really made my life much easier.
Right now, I have been trying to specialise a sentence embedding model from HuggingFace
and I was doubtful about what the various losses actually compute.

Given two sentences [x, y], the model I am using was trained to give cos_sim(x, y) = 1.

For now, I have been trying to use:

losses.ContrastiveLoss(distance = CosineSimilarity())
losses.TripletMarginLoss(distance = CosineSimilarity())

Is that correct? Does distance = CosineSimilarity() refer to the cosine_distance = (1-cosine(x, y))/2?

I am asking because the loss for each batch I have obtaining very low loss.

Thank you!

Best.

Giorgio

**Comment by KevinMusgrave** [comment]
Your usage is correct.

> Is that correct? Does distance = CosineSimilarity() refer to the cosine_distance = (1-cosine(x, y))/2?

It's a bit confusing, but actually `distance = CosineSimilarity()` is just cosine similarity, i.e. 1 is most similar and -1 is most dissimilar. So `distance` is used either as an actual distance or as a similarity.

> I am asking because the loss for each batch I have obtaining very low loss.

You might want to change the margins, especially for ContrastiveLoss. The default settings are `pos_margin=0, neg_margin=1` which doesn't make sense for cosine similartiy. You could try `pos_margin=1, neg_margin=0` instead.

**Comment by GiorgioBarnabo** [comment]
Dear Kevin,

thank you a lot for your answer. I have been experimenting a bit with your implementation of the contrastive loss and I still have some doubts. In a nutshell, I have been trying to fine-tune a sentence embedding model from HuggingFace with the following loss config: 

losses.ContrastiveLoss(distance = CosineSimilarity(), pos_margin=1, neg_margin=-1)

Nonetheless, regardless of the batch size, the loss is always between 0 and 2, as it was the average loss over all possible pairs inside the batch. Worse than that, apparently le loss start at around 1,85 and then stabilises at 2 - that is, its worst possible value.

Could you please explain me how the pairs are formed and how their loss is summed?

Thanks. Best,

Giorgio

**Comment by KevinMusgrave** [comment]
All positive pairs and negative pairs are extracted from the batch using the provided labels. The loss is computed for all pairs and the average of the **non zero** losses is returned. So the loss may increase, but the number of pairs contributing to the loss may be decreasing. 

It might be more informative to log the number of pairs that have non-zero loss:

```python
from pytorch_metric_learning.utils import common_functions
common_functions.COLLECT_STATS = True

loss_fn = ContrastiveLoss()
reducers = loss_fn.reducer.reducers

loss = loss_fn(emb, labels)
print(reducers["pos_loss"].pos_pairs_past_filter)
print(reducers["neg_loss"].neg_pairs_past_filter)
```

**Comment by GiorgioBarnabo** [comment]
Hi Kevin,

thank you very much for your quick answer. I will do it immediately. Nonetheless, O was thinking that, for a batch with n classes and m example per classes, the negative pairs are always much more than the positive ones. Is there a way to mitigate this problem?

Best,

Giorgio

**Comment by KevinMusgrave** [comment]
At the moment I can't think of an easy way to do that. Actually it shouldn't be a problem because the final loss: `(average(pos_losses) + average(neg_losses)) / 2`, i.e. the positive and negative losses are averaged independently.

**Comment by KevinMusgrave** [comment]
You could try lowering the learning rate. I'm not sure what else would cause that.

**Comment by icsd13152** [comment]
> Ahhh, I see. Still there are a new strange things happening. Apparently the optimisation pushes all the cosine similarities to 1, regardless of the label. It follows that the loss is 0 for the positive examples, and 2 for the negative ones. Nonetheless, the model gets stuck and does not reduce the loss on the negative samples :(

Hi @GiorgioBarnabo  Have you managed to found what caused your issue? I am stuck with those issues too... I have tried many values for learning rate and many values for margins.

Any answer it would be helpfull.

Thanks in advanced
